2 Analyzing GMO

In [297]:
import numpy as np
import pandas as pd
import portfolio_management_helper as pmh
import statsmodels.api as sm

rf_rts = pd.read_excel("../data/gmo_data.xlsx",sheet_name="risk-free rate").set_index("date")
gmo_rts = pd.read_excel("../data/gmo_data.xlsx",sheet_name="total returns").set_index("date")
excess_rts = gmo_rts.subtract(rf_rts["TBill 3M"]/12,axis=0)

excess_rts

SPY   GMWAX   GMGEX
date                              
1996-12-31 -0.0276 -0.0264 -0.0173
1997-01-31  0.0575  0.0104  0.0302
1997-02-28  0.0052  0.0179  0.0084
1997-03-31 -0.0502 -0.0196 -0.0209
1997-04-30  0.0600 -0.0111 -0.0044
...            ...     ...     ...
2024-06-28  0.0308 -0.0117 -0.0175
2024-07-31  0.0077  0.0260  0.0303
2024-08-30  0.0191  0.0104  0.0153
2024-09-30  0.0172  0.0086  0.0115
2024-10-31 -0.0127 -0.0407 -0.0433

[335 rows x 3 columns]

In [298]:
pmh.calc_summary_statistics(
    returns=excess_rts[["GMWAX"]],
    annual_factor=12,
    provided_excess_returns=True,
    timeframes={
        "1927-2011": ["1927", "2011"],
        "2012-2024": ["2012", "2024"],
        "1927-2024": ["1927", "2024"],
    },
    keep_columns=["Annualized Mean", "Annualized Vol", "Annualized Sharpe"],
)

Annualized Mean  Annualized Vol  Annualized Sharpe
GMWAX 1927-2011           0.0464          0.1105             0.4201
GMWAX 2012-2024           0.0434          0.0949             0.4573
GMWAX 1927-2024           0.0450          0.1035             0.4352

The mean and the Sharpe both dipped from 2012 - 2024, but the volatiity improved somewhat during that time period, too. Overall, these factors are somewhat consistent through the different time frames.

In [299]:
pmh.calc_summary_statistics(
    returns=excess_rts[["GMWAX"]],
    annual_factor=12,
    provided_excess_returns=True,
    var_quantile = .05,
    timeframes={
        "1927-2011": ["1927", "2011"],
        "2012-2024": ["2012", "2024"],
        "1927-2024": ["1927", "2024"],
    },
    keep_columns=["Min", "Annualized Historical VaR", "Max Drawdown"],
)


Min  Annualized Historical VaR (5.00%)  Max Drawdown
GMWAX 1927-2011 -0.1492                            -0.1524       -0.3065
GMWAX 2012-2024 -0.1150                            -0.1415       -0.2256
GMWAX 1927-2024 -0.1492                            -0.1433       -0.3065

The GMWAX does seem to have fairly high tail risk, considering that the max drawdown is nearly 30%, and the 5th percentile VaR is 14%. This does improv a bit in 2012-2024 versus 1927-2011.

In [300]:
pmh.calc_regression(
    y = excess_rts[["GMWAX"]],
    X = excess_rts[["SPY"]],
    annual_factor=12,
    timeframes={
        "1927-2011": ["1927", "2011"],
        "2012-2024": ["2012", "2024"],
        "1927-2024": ["1927", "2024"],
    },
    keep_columns=["Annualized Alpha", "Beta", "R-Squared"],
)

"calc_regression" assumes excess returns to calculate Information and Treynor Ratios


Annualized Alpha  R-Squared  SPY Beta
GMWAX 1927-2011            0.0270     0.6487    0.5421
GMWAX 2012-2024           -0.0340     0.7487    0.5818
GMWAX 1927-2024            0.0006     0.6802    0.5526

The GMWAX is a low beta strategy, as its beta is under 1 in all the time series. However, the beta does grow a bit higher from 2012 onwards. 

The GMWAX does provide alpha on average, although since 2012 that was deteorated significantly, where it has since actually since provided a negative alpha.

In [301]:
pmh.calc_summary_statistics(
    returns=excess_rts[["GMGEX"]],
    annual_factor=12,
    provided_excess_returns=True,
    timeframes={
        "1927-2011": ["1927", "2011"],
        "2012-2024": ["2012", "2024"],
        "1927-2024": ["1927", "2024"],
    },
    keep_columns=["Annualized Mean", "Annualized Vol", "Annualized Sharpe"],
)

Annualized Mean  Annualized Vol  Annualized Sharpe
GMGEX 1927-2011          -0.0038          0.1473            -0.0260
GMGEX 2012-2024           0.0013          0.2356             0.0056
GMGEX 1927-2024          -0.0015          0.1926            -0.0076

In [302]:
pmh.calc_summary_statistics(
    returns=excess_rts[["GMGEX"]],
    annual_factor=12,
    provided_excess_returns=True,
    var_quantile = .05,
    timeframes={
        "1927-2011": ["1927", "2011"],
        "2012-2024": ["2012", "2024"],
        "1927-2024": ["1927", "2024"],
    },
    keep_columns=["Min", "Annualized Historical VaR", "Max Drawdown"],
)

Min  Annualized Historical VaR (5.00%)  Max Drawdown
GMGEX 1927-2011 -0.1516                            -0.2851       -0.5640
GMGEX 2012-2024 -0.6589                            -0.2357       -0.7383
GMGEX 1927-2024 -0.6589                            -0.2640       -0.7681

In [303]:
pmh.calc_regression(
    y = excess_rts[["GMGEX"]],
    X = excess_rts[["SPY"]],
    annual_factor=12,
    timeframes={
        "1927-2011": ["1927", "2011"],
        "2012-2024": ["2012", "2024"],
        "1927-2024": ["1927", "2024"],
    },
    keep_columns=["Annualized Alpha", "Beta", "R-Squared"],
)

"calc_regression" assumes excess returns to calculate Information and Treynor Ratios


Annualized Alpha  R-Squared  SPY Beta
GMGEX 1927-2011           -0.0312     0.7259    0.7642
GMGEX 2012-2024           -0.1102     0.2525    0.8381
GMGEX 1927-2024           -0.0648     0.3979    0.7867

The GMGEX performs worse across the board: it has worse Sharpe, worse drawdowns, and offers less alpha/higher correlation with SPY.

3 Forecast Regressions

In [304]:
signals = pd.read_excel("../data/gmo_data.xlsx",sheet_name="signals").set_index("date")
signals.tail()

SPX DVD YLD  SPX P/E  TNote 10YR
date                                        
2024-06-28       1.3271  25.5386      4.3961
2024-07-31       1.3205  25.7965      4.0296
2024-08-30       1.2990  25.5680      3.9034
2024-09-30       1.2789  26.1183      3.7809
2024-10-31       1.3046  25.8491      4.2844

In [305]:
pmh.calc_regression(
    y = excess_rts[["SPY"]],
    X = signals[["SPX DVD YLD"]].shift(-1),
    annual_factor=12,
    keep_columns=["Annualized Alpha", "Beta", "R-Squared"],
)

"calc_regression" assumes excess returns to calculate Information and Treynor Ratios


Annualized Alpha  R-Squared  SPX DVD YLD Beta
SPY            0.3540     0.0127           -0.0125

In [306]:
pmh.calc_regression(
    y = excess_rts[["SPY"]],
    X = signals[["SPX P/E"]].shift(-1),
    annual_factor=12,
    keep_columns=["Annualized Alpha", "Beta", "R-Squared"],
)

"calc_regression" assumes excess returns to calculate Information and Treynor Ratios


Annualized Alpha  R-Squared  SPX P/E Beta
SPY           -0.2333     0.0165        0.0013

In [307]:
pmh.calc_regression(
    y = excess_rts[["SPY"]],
    X = signals[["TNote 10YR"]].shift(-1),
    annual_factor=12,
    keep_columns=["Annualized Alpha", "Beta", "R-Squared"],
)

"calc_regression" assumes excess returns to calculate Information and Treynor Ratios


Annualized Alpha  R-Squared  TNote 10YR Beta
SPY            0.1293     0.0014          -0.0011

In [308]:
pmh.calc_regression(
    y = excess_rts[["SPY"]],
    X = signals[["SPX DVD YLD","SPX P/E","TNote 10YR"]].shift(-1),
    annual_factor=12,
    keep_columns=["Annualized Alpha", "Beta", "R-Squared"],
)

"calc_regression" assumes excess returns to calculate Information and Treynor Ratios


Annualized Alpha  R-Squared  SPX DVD YLD Beta  SPX P/E Beta  \
SPY            0.2438     0.0255           -0.0107        0.0009   

     TNote 10YR Beta  
SPY          -0.0032

In [309]:
variables = [
    ("SPX DVD YLD", "DP Returns"),
    ("SPX P/E", "PE Returns"),
    ("TNote 10YR", "TNote Returns")
]

returnsdf = pd.DataFrame(index=gmo_rts.index)

for var, label in variables:
    strat = pmh.calc_regression(
        y=excess_rts[["SPY"]],
        X=signals[[var]].shift(-1),
        annual_factor=12,
        return_model=True,
    )
    
    beta = strat.params[var]
    alpha = strat.params["const"]
    
    fcst = alpha + beta * signals[var]
    fcstweighted = 100 * fcst
    fcstreturns = fcstweighted * gmo_rts["SPY"].shift(-1)
    
    returnsdf[label] = fcstreturns

returnsdf = returnsdf[:-1]
returnsdf


"calc_regression" assumes excess returns to calculate Information and Treynor Ratios
"calc_regression" assumes excess returns to calculate Information and Treynor Ratios
"calc_regression" assumes excess returns to calculate Information and Treynor Ratios


DP Returns  PE Returns  TNote Returns
date                                             
1996-12-31      0.0301      0.0410         0.0213
1997-01-31      0.0061      0.0079         0.0032
1997-02-28     -0.0288     -0.0383        -0.0150
1997-03-31      0.0331      0.0323         0.0186
1997-04-30      0.0404      0.0412         0.0196
...                ...         ...            ...
2024-05-31      0.0438      0.0466         0.0199
2024-06-28      0.0156      0.0174         0.0070
2024-07-31      0.0302      0.0344         0.0144
2024-08-30      0.0278      0.0303         0.0133
2024-09-30     -0.0120     -0.0135        -0.0058

[334 rows x 3 columns]

In [310]:
pmh.calc_summary_statistics(
    returns=returnsdf,
    annual_factor=12,
    provided_excess_returns=True,
    var_quantile = .05,
    keep_columns=["Annualized Mean","Annualized Vol","Annualized Sharpe","Max Drawdown"],
)


Annualized Mean  Annualized Vol  Annualized Sharpe  \
DP Returns              0.0436          0.1384             0.3148   
PE Returns              0.0505          0.1354             0.3734   
TNote Returns           0.0755          0.1078             0.7003   

               Max Drawdown  
DP Returns          -0.5356  
PE Returns          -0.5499  
TNote Returns       -0.3710

In [311]:
returnStrats = ["DP Returns", "PE Returns", "TNote Returns"]
results_list = []

for strat in returnStrats:
    regression_result = pmh.calc_regression(
        y=returnsdf[[strat]],
        X=excess_rts[["SPY"]],
        annual_factor=12,
        calc_treynor_info_ratios=True,
        keep_columns=["Annualized Alpha", "Beta", "Information Ratio"],
    )

    alpha = regression_result["Annualized Alpha"].iloc[0]
    beta = regression_result["SPY Beta"].iloc[0]
    info_ratio = regression_result["Information Ratio"].iloc[0]

    results_list.append({
        "Strategy": strat,
        "Annualized Alpha": alpha,
        "Beta": beta,
        "Information Ratio": info_ratio
    })

regression_metrics = pd.DataFrame(results_list)
regression_metrics.set_index("Strategy", inplace=True)

regression_metrics


"calc_regression" assumes excess returns to calculate Information and Treynor Ratios
y has lenght 334 and X has lenght 335. Joining y and X by index...
"calc_regression" assumes excess returns to calculate Information and Treynor Ratios
y has lenght 334 and X has lenght 335. Joining y and X by index...
"calc_regression" assumes excess returns to calculate Information and Treynor Ratios
y has lenght 334 and X has lenght 335. Joining y and X by index...


Annualized Alpha    Beta  Information Ratio
Strategy                                                  
DP Returns               0.0481 -0.0565             0.1007
PE Returns               0.0524 -0.0233             0.1119
TNote Returns            0.0754  0.0014             0.2019

In [312]:
pmh.calc_summary_statistics(
    returns=returnsdf,
    annual_factor=12,
    provided_excess_returns=True,
    var_quantile = .05,
    keep_columns=["Historical VaR"],
)

Historical VaR (5.00%)  Annualized Historical VaR (5.00%)
DP Returns                    -0.0719                            -0.2491
PE Returns                    -0.0612                            -0.2120
TNote Returns                 -0.0505                            -0.1749

In [313]:
pmh.calc_summary_statistics(
    returns=excess_rts[["SPY","GMWAX"]],
    annual_factor=12,
    provided_excess_returns=True,
    var_quantile = .05,
    keep_columns=["Historical VaR"],
)

Historical VaR (5.00%)  Annualized Historical VaR (5.00%)
SPY                   -0.0797                            -0.2760
GMWAX                 -0.0414                            -0.1433

In [314]:
returnsdfexcess = returnsdf.subtract(rf_rts["TBill 3M"][:-1]/12,axis=0)

In [315]:
filtered_rts = returnsdfexcess[(returnsdfexcess.index >= '2001-01-31') & (returnsdfexcess.index <= '2011-12-31')]
filtered_rts.mean()


DP Returns      -0.0040
PE Returns      -0.0051
TNote Returns    0.0006
dtype: float64

The dividend-price and price to earnings portfolio underperform the risk free rate, while the tnote overpeforms, but just slightly.

In [316]:
negative_counts = (returnsdfexcess < 0).sum()
positive_counts = (returnsdfexcess > 0).sum()

counts = pd.DataFrame({
    'Negative Returns': negative_counts,
    'Positive Returns': positive_counts
})

print(counts)

               Negative Returns  Positive Returns
DP Returns                  129               205
PE Returns                  148               186
TNote Returns               128               206


In our reression, we found that the DP and PE strategies have negative betas, while the TNote strategy has a positive beta. Therfore, any period where the DP and PE have positive returns results in a negative risk premium, vice versa for the TNotes. Ergo, DP has 205 negative risk premium periods, PE has 186 negative risk premium periods, and Tnote has 128 negative risk premium periods.

Compared to GMWAX, the dynamic strategies have less systematic risk, as their betas to SPY are lower, implying less correlation with the market. However, they do have higher unsystematic risk, as their VaRs are all higher.

Out-of-Sample Forecasting

In [317]:
periods=60

In [318]:
DP_IS_model = pmh.calc_regression(
    y = excess_rts[["SPY"]][0:t],
    X = signals[["SPX DVD YLD"]].shift(-1)[0:t],
    annual_factor=12,
    return_model= True,
)

"calc_regression" assumes excess returns to calculate Information and Treynor Ratios


In [319]:
PE_IS_model = pmh.calc_regression(
    y = excess_rts[["SPY"]][0:t],
    X = signals[["SPX P/E"]].shift(-1)[0:t],
    annual_factor=12,
    return_model= True,
)

"calc_regression" assumes excess returns to calculate Information and Treynor Ratios


In [320]:
DPerrors = []
t = periods

for t in range(t,len(signals)-1):
    out_sample_X = signals[["SPX DVD YLD"]].shift(-1)[t+1:t+2]
    out_sample_X_with_const = sm.add_constant(out_sample_X, has_constant="add")
    DP_out_sample_predictions = DP_IS_model.predict(out_sample_X_with_const)
    error = excess_rts[["SPY"]].iloc[t].squeeze() - DP_out_sample_predictions.iloc[0]
    DPerrors.append(error)

In [321]:
PEerrors = []
t = periods

for t in range(t,len(signals)-1):
    out_sample_X = signals[["SPX P/E"]].shift(-1)[t+1:t+2]
    out_sample_X_with_const = sm.add_constant(out_sample_X, has_constant="add")
    DP_out_sample_predictions = DP_IS_model.predict(out_sample_X_with_const)
    error = excess_rts[["SPY"]].iloc[t].squeeze() - DP_out_sample_predictions.iloc[0]
    DPerrors.append(error)